File: building-a-wallet/python/step1_fetch_state.py
import sys
from solana.rpc.api import Client
from solana.rpc.types import TokenAccountOpts
from solders.pubkey import Pubkey

ENDPOINT = "https://your-endpoint.rpcpool.com"
TOKEN = "your-token"
TOKEN_PROGRAM_ID = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"

def get_wallet_state(wallet_address: str):
    client = Client(f"{ENDPOINT}/{TOKEN}")
    pubkey = Pubkey.from_string(wallet_address)

    # SOL balance
    resp = client.get_balance(pubkey)
    print(f"SOL: {resp.value / 1_000_000_000:.9f}")

    # SPL token accounts
    opts = TokenAccountOpts(program_id=Pubkey.from_string(TOKEN_PROGRAM_ID))
    token_resp = client.get_token_accounts_by_owner_json_parsed(pubkey, opts)
    for account in token_resp.value:
        info = account.account.data.parsed["info"]
        mint = info["mint"]
        amount = info["tokenAmount"]["uiAmount"]
        print(f"Token {mint}: {amount}")

wallet = sys.argv[1] if len(sys.argv) > 1 else "YourWalletAddressHere"
get_wallet_state(wallet)

File: building-a-wallet/typescript/step1-fetch-state.ts
import {
  Connection,
  PublicKey,
  LAMPORTS_PER_SOL,
} from "@solana/web3.js";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";

const ENDPOINT = "https://your-endpoint.rpcpool.com";
const TOKEN = "your-token";

async function getWalletState(walletAddress: string) {
  const connection = new Connection(`${ENDPOINT}/${TOKEN}`);
  const pubkey = new PublicKey(walletAddress);

  // SOL balance
  const lamports = await connection.getBalance(pubkey);
  console.log(`SOL: ${lamports / LAMPORTS_PER_SOL}`);

  // All SPL token accounts owned by this wallet
  const { value: tokenAccounts } = await connection.getParsedTokenAccountsByOwner(
    pubkey,
    { programId: TOKEN_PROGRAM_ID }
  );

  for (const { account } of tokenAccounts) {
    const info = account.data.parsed.info;
    const mint: string = info.mint;
    const amount: number = info.tokenAmount.uiAmount;
    console.log(`Token ${mint}: ${amount}`);
  }
}

const wallet = process.argv[2] ?? "YourWalletAddressHere";
getWalletState(wallet).catch(console.error);
